aws/resource_aws_codepipeline_webhook.go

```diff
@@ -0,0 +1,336 @@
+package aws
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/codepipeline"
+)
+
+func resourceAwsCodePipelineWebhook() *schema.Resource {
+	return &schema.Resource{
+		Create:        resourceAwsCodePipelineWebhookCreate,
+		Read:          resourceAwsCodePipelineWebhookRead,
+		Update:        nil,
+		Delete:        resourceAwsCodePipelineWebhookDelete,
+		SchemaVersion: 1,
+
+		Schema: map[string]*schema.Schema{
+			"authentication": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					codepipeline.WebhookAuthenticationTypeGithubHmac,
+					codepipeline.WebhookAuthenticationTypeIp,
+					codepipeline.WebhookAuthenticationTypeUnauthenticated,
+				}, false),
+			},
+			"authentication_configuration": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				MinItems: 1,
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"secret_token": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"allowed_ip_range": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.CIDRNetwork(0, 32),
+						},
+					},
+				},
+			},
+			"filter": {
+				Type:     schema.TypeSet,
+				ForceNew: true,
+				Required: true,
+				MinItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"json_path": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						"match_equals": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"target_action": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"target_pipeline": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+		},
+	}
+}
+
+func extractCodePipelineWebhookAttr(d *schema.ResourceData, attr string) (map[string]interface{}, error) {
```

Comment:
We prefer to allow the schema and any API errors to dictate this behavior rather than implementing additional error handling in the resource functions.  `authentication_configuration` is marked as `Required: true` and we can allow the API (or the handling in `extractCodePipelineWebhookAuthConfig`) to return errors if both `Optional` nested arguments are omitted.